plugin/metrics/README.md

```diff
@@ -39,6 +39,13 @@ name "dropped" (without a closing dot - this is never a valid domain name).
 
 This plugin can only be used once per Server Block.
 
+## Health
+
+This plugin implements dynamic health checking. 
+Prometheus plugin is considered unhealthy if the HTTP server instance was not successfully started.
```

Comment:
so, we will not fail binding to the port? That was an earlier plan, no? Did we decide we could not do that ?

Reply:
as you started to say in the issue #1868, the problem to fail is that if we fail to start the new server, the CoreDNS will continue service on the old server .. but the listener is in unknown state (most likely started to stop but not finished). Which means Metrics most likely stopped to serve information on either new or old server.

It does not help to automate the monitoring of CoreDNS, unless we completely stop CoreDNS.

Adding Metrics to Health and inform when metrics is unhealthy allow any monitor to take a decision for stop/start CoreDNS

Reply:
I keep finding this odd. We should probably crash the startup if we fail to bind, i.e. not fail the reload, crash the entire process. Seems simplest and avoid a whole bunch of code to work around it. I do wonder however if log.Fatalf would crash everything or just the reloading process.

I also don't understand why your monitoring doesn't see the monitoring endpoint disappearing and needs the health endpoint to do something

Reply:
# Short

- **I can remove the healther implementation. I do not need it. But I would like to keep the move out of global var - which also fix #1868 **
- **I would not like to have to stop the process if Metrics cannot listen. At least for few weeks. That would block IB that try to go to production right now.**

# Longer reply 

I think this PR contains several changes, and maybe I should go one by one:

## 1- issue #1868
it is the impact of Reload on the Metrics plugin that is using uniq resources (the Listeners). 
IMO the root cause is the usage of global var to save information : we cannot cleanup this information during a reload. That get worse when the Reload fails for some reason and we expect to go back to initial server.
=> Her, I move the uniqAddr info from global var, to Instance related storage. But still share by all Metrics structures used during that instance.


## 2- Healther and what to do when the listener (uniq resource) cannot start.
I was going toward Health plugin for consistency : it should warn when CoreDNS service is not correct.
But I agree that EXIT the process is another valid solution.
NOTE: having the monitoring check the Metrics endpoint along with Health endpoint can work too .. but that is inconsistent with Health definition.

## 3- the size of the change ... hum here I am not following too much. 
The point is to have the plugin working correctly.
Here I tried to fix the following issues I see in the current code:
- move away from the global var 
- change init uniqAddr in Setup- away from the go function that we are not sure when it will run (I mean nothing ensure it will run before the Start of the server)
- set status in UniqAddr to "done" only when listener is really started. Seems sanity. But also allow to have either an external check (Health) or retry (if we want ...)
- mutex on UniqAddr functions. is sanity. We are lucky so far this lib is used only at setup or event that are synced.

As the code is done and proposed now, isn't it better to checked-in ?

- the one that want to monitor on endpoint Metrics can still continue
- the issue #1868 is fixed with this code
- lib uniqAddr is thread-safe and dan be reused 

and one usage I may have (depends internal Infoblox) is ability to retry the Start of the Metrics from within an external plugin which is doing CLI+monitoring.


## What I expect from metrics for IB usage
- fix of #1868 (after several reload, Metrics lost ability to stop/start the listener according to Corefile)
- ability to know it Metrics failed to launch the Listener
- ability to retry (ability to know is similar to Healther). and the retry is described into RepairHealth in UT.

NOTE: Right now, the EXIT in Metrics in case of fail to listen would hurt IB use case. 
They need to move up to Go 1.10. I am pushing on it. But not done yet.